### PR TITLE
Panel export: change IDs #841

### LINKED ISSUE #841: Panel export: change IDs
Exported panel can't be imported within the same Grafana instance, there will be errors because of IDs collisions.

We should replace all IDs with placeholders during the exporting process.

### PR DESCRIPTION
**Pull request by rozetko**
Closes #841 

This PR changes panel importing process:  when panel is being imported, IDs of analytic units / caches / segments / detection spans are changed to new ones

## Disclaimer
Abbreviations in this PR:
- AU -> analytic unit
- AUC -> analytic unit cache
- S -> segment
- DS -> detection span

## New import process:
- remove IDs from AU / AUC / S / DS while exporting
- import AU, database generates new IDs for them
- import AUC / S / DS with new analytic unit IDs

### CODE REVIEW

**Comment by rozetko** [comment]
@jonyrock-back 
Please, review it

File: server/src/services/grafana_service.ts [hunk 1]
```diff
@@ -29,13 +31,51 @@ export async function importPanel(
   panelTemplate: GrafanaPanelTemplate,
   variables: GrafanaTemplateVariables
 ): Promise<void> {
+  const oldAnalyticUnitIds = panelTemplate.analyticUnits.map(analyticUnit => analyticUnit._id);
+
   panelTemplate.analyticUnits.forEach(analyticUnit => {
+    analyticUnit._id = undefined;
```
Comment (jonyrock-back) [inline server/src/services/grafana_service.ts]: do you need to explicity set `analyticUnit._id = undefined`? Maybe you need `delete analyticUnit._id` here?

Comment (jonyrock-back) [reply server/src/services/grafana_service.ts]: I personally think that you should warn if you find id in analytic unit, it not suppose to be here, right?
These id's should be removed on export, no?

File: server/src/services/grafana_service.ts [hunk 2]
```diff
@@ -29,13 +31,51 @@ export async function importPanel(
...
-  await AnalyticUnit.insertMany(panelTemplate.analyticUnits);
-  await AnalyticUnitCache.insertMany(panelTemplate.caches);
-  await Segment.insertMany(panelTemplate.segments);
-  await DetectionSpan.insertMany(panelTemplate.detectionSpans);
+
+  const newAnalyticUnitIds = await AnalyticUnit.insertMany(panelTemplate.analyticUnits);
+
+  if(newAnalyticUnitIds.length !== oldAnalyticUnitIds.length) {
+    throw new Error(`
+      Something went wrong while inserting analytic units:
```
Comment (jonyrock-back) [inline server/src/services/grafana_service.ts]: Please remove this line. This line would be relevant in outer catch block

File: server/src/services/grafana_service.ts [hunk 3]
```diff
@@ -29,13 +31,51 @@ export async function importPanel(
...
+    })
+  );
+  const newDetectionSpans = panelTemplate.detectionSpans.map(
+    span => ({
+      ...span,
+      analyticUnitId: oldToNewAnalyticUnitIdsMapping.get(span.analyticUnitId)
+    })
+  );
+
+  await Promise.all([
```
Comment (jonyrock-back) [inline server/src/services/grafana_service.ts]: what if one of these Promises breaks? You will partially insert data?
Yeah, I know it's a pain to make inserting right

File: server/src/services/grafana_service.ts (hunk 1 above)
Comment (rozetko) [reply server/src/services/grafana_service.ts]: Now it's not error: we don't remove IDs on export, but maybe this behavior should be changed

File: server/src/services/grafana_service.ts (hunk 3 above)
Comment (rozetko) [reply server/src/services/grafana_service.ts]: Yes, we should make some kind of transactions here

File: server/src/models/detection_model.ts [hunk 4]
```diff
@@ -57,6 +57,14 @@ export class DetectionSpan {
     };
   }
 
+  public toTemplate(): any {
+    return {
+      ...this.toObject(),
+      _id: undefined,
```
Comment (jonyrock-back) [inline server/src/models/detection_model.ts]: same thing with `delete` here. Why do you set it to `undefined`?

Comment (rozetko) [reply server/src/models/detection_model.ts]: it's just more convenient here

**Pull request merged**